ntirpc/rpc/clnt.h

```diff
@@ -281,13 +261,79 @@ struct rpc_timers {
 #define CLSET_PUSH_TIMOD 17	/* push timod if not already present */
 #define CLSET_POP_TIMOD  18	/* pop timod */
 
-/*
- * void
- * CLNT_DESTROY(rh);
- *  CLIENT *rh;
+/* Protect a CLIENT with a CLNT_REF for each call or request.
  */
-#define CLNT_DESTROY(rh) ((*(rh)->cl_ops->cl_destroy)(rh))
-#define clnt_destroy(rh) ((*(rh)->cl_ops->cl_destroy)(rh))
+static inline void clnt_ref_it(CLIENT *clnt, uint32_t flags,
+			       const char *tag, const int line)
+{
+	uint32_t refs = atomic_inc_uint32_t(&clnt->cl_refcnt);
+
+	if (flags & CLNT_REF_FLAG_LOCKED)  {
+		/* unlock before warning trace */
+		mutex_unlock(&clnt->cl_lock);
+	}
+	__warnx(TIRPC_DEBUG_FLAG_REFCNT, "%s: %p %" PRIu32 " @%s:%d",
+		__func__, clnt, refs, tag, line);
+}
+#define CLNT_REF(clnt, flags)						\
+	clnt_ref_it(clnt, flags, __func__, __LINE__)
+
+static inline void clnt_release_it(CLIENT *clnt, uint32_t flags,
+				   const char *tag, const int line)
+{
+	uint32_t refs = atomic_dec_uint32_t(&clnt->cl_refcnt);
+	uint16_t cl_flags;
+
+	if (flags & CLNT_RELEASE_FLAG_LOCKED) {
+		/* unlock before warning trace */
+		mutex_unlock(&clnt->cl_lock);
+	}
+	__warnx(TIRPC_DEBUG_FLAG_REFCNT, "%s: %p %" PRIu32 " @%s:%d",
+		__func__, clnt, refs, tag, line);
+
+	if (likely(refs > 0)) {
+		/* normal case */
+		return;
+	}
+
+	/* enforce once-only semantic, trace others */
+	cl_flags = atomic_postset_uint16_t_bits(&clnt->cl_flags,
+						CLNT_FLAG_RELEASING);
+
+	if (cl_flags & CLNT_FLAG_RELEASING) {
+		__warnx(TIRPC_DEBUG_FLAG_ERROR,
+			"%s: %p WARNING! already destroying! @%s:%d",
+			__func__, clnt, tag, line);
+		return;
+	}
+
+	/* Releasing last reference */
+	(*(clnt)->cl_ops->cl_destroy)(clnt);
+}
+#define CLNT_RELEASE(clnt, flags)					\
+	clnt_release_it(clnt, flags, __func__, __LINE__)
+
+/* CLNT_DESTROY is CLNT_RELEASE with once-only semantics.  Also, idempotent
+ * CLNT_FLAG_DESTROYED indicates that more references should not be taken.
+ */
+static inline void clnt_destroy_it(CLIENT *clnt,
```

Comment:
clnt_release already has a flag making it once-only.  What's the benifit of clnt_destroy?